components/remote_settings/src/client.rs

```diff
@@ -18,6 +19,205 @@ const HEADER_BACKOFF: &str = "Backoff";
 const HEADER_ETAG: &str = "ETag";
 const HEADER_RETRY_AFTER: &str = "Retry-After";
 
+/// Internal Remote settings client API
+pub struct RemoteSettingsClient {
+    // This is immutable, so it can be outside the mutex
+    collection_name: String,
+    inner: Mutex<RemoteSettingsClientInner>,
+}
+
+struct RemoteSettingsClientInner {
+    storage: Storage,
+    api_client: ApiClient,
+}
+
+impl RemoteSettingsClient {
+    pub fn new(
+        server_url: Url,
+        bucket_name: String,
+        collection_name: String,
+        storage: Storage,
+    ) -> Result<Self> {
+        let api_client = ApiClient::new(server_url, &bucket_name, &collection_name)?;
+        Ok(Self {
+            collection_name,
+            inner: Mutex::new(RemoteSettingsClientInner {
+                storage,
+                api_client,
+            }),
+        })
+    }
+
+    pub fn collection_name(&self) -> &str {
+        &self.collection_name
+    }
+
+    pub fn get_records(&self, sync_if_empty: bool) -> Result<Option<Vec<RemoteSettingsRecord>>> {
+        let mut inner = self.inner.lock();
+        let base_url = inner.api_client.base_url();
+
+        let cached_records = inner.storage.get_records(&base_url)?;
+        if cached_records.is_some() || !sync_if_empty {
+            return Ok(cached_records);
+        }
+
+        let records = inner.api_client.get_records(None)?;
+        inner.storage.set_records(&base_url, &records)?;
+        Ok(Some(records))
+    }
+
+    pub fn sync(&self) -> Result<()> {
+        let mut inner = self.inner.lock();
+        let base_url = inner.api_client.base_url();
+        let mtime = inner.storage.get_last_modified_timestamp(&base_url)?;
+        let records = inner.api_client.get_records(mtime)?;
+        inner.storage.set_records(&base_url, &records)
+    }
+
+    /// Downloads an attachment from [attachment_location]. NOTE: there are no guarantees about a
+    /// maximum size, so use care when fetching potentially large attachments.
+    pub fn get_attachment(&self, attachment_location: &str) -> Result<Vec<u8>> {
+        self.inner
+            .lock()
+            .api_client
+            .get_attachment(attachment_location)
+    }
+
+    pub fn update_config(&self, server_url: Url, bucket_name: String) -> Result<()> {
+        let mut inner = self.inner.lock();
+        inner.api_client = ApiClient::new(server_url, &bucket_name, &self.collection_name)?;
+        inner.storage.empty()
+    }
+}
+
+/// Client for Remote settings API requests
+struct ApiClient {
```

Comment:
I'm never sure if I should write unit tests for these kinds of classes.  On the one hand, unit tests are great.  On the other hand, testing this would involve a bunch of mocking and I often feel like the tests end up mirroring the implementation in a way that doesn't actually test anything.  What do others think?

Another option would be e2e tests.  The `cargo rs` client provides a very rudimentary way to kind of do that.  Maybe we could extend that into a real e2e test suite.

Reply:
I think we could spin up a basic test server which returns a JSON, and then test if we can store and fetch records? I am not sure if we even need a e2e test suite for this (could spin up a server, and shut it down via a `oneshot` channel after the test finish for example. But maybe we want something more...fancy.

Reply:
I guess I disagree with myself, because I ended up writing some mock tests.  I'm still not sure how useful they are but it's not all that much code, so I figured it was worth it.